Exports IsRuleEngineOff

M4tteoP opened the pull request:
Follows https://github.com/corazawaf/coraza/pull/499, Just like `RequestBodyAccess` and `ResponseBodyAccess`, connectors could benefit from retrieving the current status of the engine. I also added a note to explicitly tell that it is the **current** status per-transaction basis in order to be aware of using it without making wrong assumptions.

jptosso reviewed (approved):
LGTM!

jcchavezs reviewed (changes requested):
Sometimes it is easy to get a too broad API because of this kind of changes where we expose an internal state and expect clients to deal with the expected behaviour so before we merge this I would expect to exercise this API in the `http` package and prove what can every value do for us.

jcchavezs commented:
Also if finally export this we should document what are the expectations for connectors as for example, `DetectionOnly` means analysis but no interruptions whereas `Off` means no analysis at all and while user might not notice, this is different in terms of audit logs.

M4tteoP commented:
Thanks @jcchavezs for the review. I see your concern. I tried to properly document the exported methods and the expected use cases. Happy to keep the conversation going, and eventually revert the changes if we feel that there is too much risk of being misused.
My main use case would be https://github.com/corazawaf/coraza-proxy-wasm/issues/68

jcchavezs commented on http/middleware.go:
```diff
@@ -89,6 +89,18 @@ func processRequest(tx types.Transaction, req *http.Request) (*types.Interruptio
+	// It is performed here as a matter of example. It is recommended to avoid doing it
+	// if not strictly needed for server/proxy side actions.
+	switch {
+	case tx.RuleEngineStatus() == types.RuleEngineOff:
```
I think if `tx.RuleEngineStatus() == types.RuleEngineOff` I would avoid the wrapping at all because not rules are evaluated and hence nothing different than the wrapped handler should be executed. Probably early termination in https://github.com/corazawaf/coraza/blob/v3/dev/http/middleware.go#L106 and not considering that case in here.

anuraaga replied:
Yeah - unless gating at a higher level, there doesn't seem to be a reason to add these branches since `tx.ProcessRequestBody` itself should be effectively a no-op already.

jcchavezs commented:
I think there are more tweaks that could be applied to middlewares/interceptors based on this. My main concern here would be: do we want to expose `RuleEngineStatus()` or just `isRuleEngineOff()`? We need a PR for proxy-wasm also exercising this API.

M4tteoP commented:
> I think there are more tweaks that could be applied to middlewares/interceptors based on this.

I agree, I'm thinking about them, but they are tricky to be integrated with the interceptor logic. 
> do we want to expose RuleEngineStatus() or just isRuleEngineOff()?

I see both values in `Off` and `DetectionOnly`. E.g. if detection only, you have to send the buffer to Coraza, but you may assume that Coraza will never require a deny action. Therefore, you may also already send data to the backend server

jcchavezs commented:
Relevant https://github.com/corazawaf/coraza/pull/505

jcchavezs commented:
I think what I am chasing here is: do we want to expose the state or just let connectors know if engine is on/off. I was playing with it in the http middleware and ended up using three states: 
- doing nothing when engine is off
- analysing the request/response body in a separate goroutine (as it won't change the response)

I guess only on/off works for proxy/wasm? how about caddy @jptosso ?

M4tteoP commented:
> analysing the request/response body in a separate goroutine (as it won't change the response)

This is Detection only status, isn't it? I'm still wondering about the corner case of a rule triggered in `detection only` with `ctl:ruleEngine=On`. In that case, we should consider that `Detection only` may still end up with disruptive action, any other early assumption may be wrong.

jcchavezs commented:
If that is the case then we definitively expose a binary value
`EngineRuleOff` or something like that as that is all conectors can know
about whitout dealing with internal logic cc @jptosso @fzipi

On Tue, 15 Nov 2022, 12:50 Matteo Pace, ***@***.***> wrote:

> analysing the request/response body in a separate goroutine (as it won't
> change the response)
>
> This is Detection only status, isn't it? I'm still wondering about the
> corner case of a rule triggered in detection only with ctl:ruleEngine=On.
> In that case, we should consider that Detection only may still end up
> with disruptive action, any other early assumption may be wrong.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/corazawaf/coraza/pull/504#issuecomment-1315199729>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXOYAQWA32XRXFZTCJU3A3WIN2HDANCNFSM6AAAAAAR5VJMHM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

fzipi commented:
I think it makes sense to expose the value or any method doing the same will do also.

jptosso commented:
@jcchavezs on Caddy the connector will still fill all variables even if the engine is off. I'm not sure if it makes sense to control that behavior from the connector side, but maybe boolean methods would be more developer-friendly, like `IsEngineEnabled()`.

jcchavezs commented:
So I guess we only need to expose `EngineRuleOff` or `IsEngineRuleOff`

M4tteoP commented:
Mostly reverted this PR.
-  The exported function is now `IsEngineRuleOff`.
- I revisited the comments about the proper usage of `RequestBodyAccessible()` and `ResponseBodyAccessible()`.

The main use case in my mind was about `DetectionOnly` status that would have permitted the connectors to send information to Coraza and at the same time let the request continue. But I came up with that it is a wrong early assumption and it would impact [at least one use case](https://github.com/corazawaf/coraza/pull/504#issuecomment-1315199729).
Sorry for the noise and many thanks for the feedback

jcchavezs commented:
LGTM

On Fri, 18 Nov 2022, 13:00 Matteo Pace, ***@***.***> wrote:

> Mostly reverted this PR.
>
>    - The exported function is now IsEngineRuleOff.
>    - I revisited the comments about the proper usage of
>    RequestBodyAccessible() and ResponseBodyAccessible().
>
> The main use case in my mind was about DetectionOnly status that would
> have permitted the connectors to send information to Coraza and at the same
> time let the request continue. But I came up with that it is a wrong early
> assumption and it would impact at least one use case
> <https://github.com/corazawaf/coraza/pull/504#issuecomment-1315199729>.
> Sorry for the noise and many thanks for the feedback
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/corazawaf/coraza/pull/504#issuecomment-1319905747>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXOYAWRTGNOLSGAPDTEW2DWI5VWBANCNFSM6AAAAAAR5VJMHM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

fzipi reviewed (approved):
LGTM also!

jcchavezs commented:
Do you mind using the condition in the HTTP middleware to return same rw when engine is off?

M4tteoP commented:
Done @jcchavezs, I hope this is what you expected. I felt was right to just keep the transaction generation thinking that Coraza is actually aware that a new request is coming, but then no rules are processed.

I also inverted "Rule" and "Engine" inside the function name being everywhere called RuleEngine (e.g. `SecRuleEngine`)

jcchavezs commented:
I think it is not correct as of now. If you return when the engine is off
then you don't even call downstream to do the response. Also worth adding a
test on this.

On Sat, 19 Nov 2022, 01:11 Matteo Pace, ***@***.***> wrote:

> Done @jcchavezs <https://github.com/jcchavezs>, I hope this is what you
> expected. I felt was right to just keep the transaction generation thinking
> that Coraza is actually aware that a new request is coming, but then no
> rules are processed.
>
> I also inverted "Rule" and "Engine" inside the function name being
> everywhere called RuleEngine (e.g. SecRuleEngine)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/corazawaf/coraza/pull/504#issuecomment-1320684920>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAXOYAVFHF5ONSBCBUVOW53WJALJZANCNFSM6AAAAAAR5VJMHM>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

M4tteoP commented:
My bad. I added the call to downstream to do the response. I'm thinking about anticipating, even more, the check to return the same response writer, but we need the transaction to check for engine off.

I added `TestHttpServerWithRuleEngineOff` decoupling `TestHttpServer` with a new helper function called `runAgainstWaf`. It adds flexibility to run different sets of requests against different Waf configs. Hope you like it 🙏
